CMakeLists.txt

```diff
@@ -354,13 +354,58 @@ if(GINKGO_BUILD_BENCHMARKS)
 endif()
 
 if(GINKGO_DEVEL_TOOLS)
-    add_custom_target(add_license
-        COMMAND ${Ginkgo_SOURCE_DIR}/dev_tools/scripts/add_license.sh
-        WORKING_DIRECTORY ${Ginkgo_SOURCE_DIR})
-    # if git-cmake-format can not build format target, do not add the dependencies
-    if(TARGET format)
+    function(enable_format_check)
+        macro(message_git_warning)
+            message(WARNING "No viable Git setup found, but GINKGO_DEVEL_TOOLS=ON is set. "
+                    "It is recommended to enable this option only with the intention of "
+                    "adding changes to Ginkgo via Git.")
+        endmacro()
+
+        find_package(Git QUIET)
+        if(NOT GIT_FOUND)
+            message_git_warning()
+            return()
+        endif()
+
+        execute_process(COMMAND ${GIT_EXECUTABLE} "rev-parse" "--git-dir"
+                        WORKING_DIRECTORY ${Ginkgo_SOURCE_DIR}
+                        RESULT_VARIABLE GINKGO_IS_GIT_DIR_ERROR)
+        if(${GINKGO_IS_GIT_DIR_ERROR})
+            message_git_warning()
+            return()
+        endif()
+
+        find_program(PRE_COMMIT pre-commit)
+        if(${PRE_COMMIT} STREQUAL "PRE_COMMIT-NOTFOUND")
+            message(FATAL_ERROR "The pre-commit command was not found. It is necessary if you want to commit changes to Ginkgo. "
+                    "If that is not the case, set GINKGO_DEVEL_TOOLS=OFF. "
+                    "Otherwise install pre-commit via pipx (or pip) using:\n"
+                    "    pipx install pre-commit")
+        endif()
+
+        execute_process(COMMAND "${PRE_COMMIT}" "install"
+                        WORKING_DIRECTORY ${Ginkgo_SOURCE_DIR}
+                        RESULT_VARIABLE pre-commit-result
+                        OUTPUT_VARIABLE pre-commit-output
+                        ERROR_VARIABLE pre-commit-error)
+        if(${pre-commit-result})
+            message(FATAL_ERROR
+                    "Failed to install the git hooks via pre-commit. Please check the error message:\n"
+                    "${pre-commit-output}\n${pre-commit-error}")
+        endif()
+
+        add_custom_target(format
+                          COMMAND bash -c "pre-commit run"
+                          WORKING_DIRECTORY ${Ginkgo_SOURCE_DIR}
+                          VERBATIM)
+
+        add_custom_target(add_license
+                          COMMAND ${Ginkgo_SOURCE_DIR}/dev_tools/scripts/add_license.sh
+                          WORKING_DIRECTORY ${Ginkgo_SOURCE_DIR})
         add_dependencies(format add_license)
-    endif()
+    endfunction()
+
+    enable_format_check()
```

Comment:
Do you use a separate function for variable scoping?

Reply:
No, because I wanted to use fast returns with the corresponding warning. Otherwise, I would get quite deep if-else nesting.